## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [53]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [54]:
# View the column names in the data
df_column = df.columns
df_column


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [55]:
# Use the describe function to gather some basic statistics

df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [56]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.loc[df['subcategory'] == 'paper products'].head()


first      last                                           job  \
8    Rebecca     Brown                                   Chiropodist   
12   Jessica     Reyes  Production designer, theatre/television/film   
13  Patricia     Gould                                      Lobbyist   
19    Shelby  Williams                           Colour technologist   
24     Glenn    Garcia                              Textile designer   

                    phone                              email  client_id  \
8        665-955-0156x926               rbrown850@lewis.info      73362   
12             5016915923           jessica.reyes@grimes.net      33615   
13  +1-694-749-5807x94968           newspapergould@walsh.biz      18145   
19       664.584.8160x389        williamstry70@armstrong.com      77184   
24           573.981.6374  determinegarcia5685@velasquez.com      33804   

    order_id  order_date  order_week  order_year         item_id     category  \
8    9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
12   6059361  2023-05-16          20        2023  BDC65078-60-7M  consumables   
13   3470698  2023-04-19          16        2023  AME10969-48-7D  consumables   
19   7573941  2023-05-29          22        2023  ZDX29223-32-0B  consumables   
24   6024379  2023-01-26           4        2023  XZZ76042-95-6X  consumables   

       subcategory  unit_price  unit_cost  unit_weight   qty  line_number  
8   paper products       83.86      59.00         0.27    49            3  
12  paper products      229.74     198.38        27.51   568            2  
13  paper products       33.72      30.08         2.01    44            0  
19  paper products      199.05     183.99         8.11   233            4  
24  paper products        5.17       3.53         6.90  2163            2

In [57]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [58]:
# For the category with the most entries,
# which subcategory had the most entries?
df['subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [59]:
# Which five clients had the most entries in the data?
client_df = df['client_id'].value_counts().head(5)


In [60]:
# Store the client ids of those top 5 clients in a list.

top5_client_list = list(df['client_id'].value_counts().head(5).index)
top5_client_list

[33615, 66037, 46820, 38378, 24741]

In [91]:
# How many total units (the qty column) did the
# client with the most entries order order?

first_client_consumption = df.loc[df['client_id'] == top5_client_list[0], 'qty'].sum()
first_client_consumption


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [84]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['unit_price_sum'] = df['unit_price']*df['qty']
df[['unit_price','qty','unit_price_sum']].head(2)


unit_price  qty  unit_price_sum
0     1096.80  105       115164.00
1       24.95   21          523.95

In [94]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
total_weight = df['unit_weight'] * df['qty']
df['total_weight']=total_weight
def cal_shipping_rate(weight):
    if weight >50:
        return weight * 7
    else :
        return weight * 10
df['shipping_price'] = df['total_weight'].apply(cal_shipping_rate)
df[['unit_price','unit_weight','qty','total_weight','shipping_price']].head(3)


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [113]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
#df['subtotal']=df['unit_price']*df['qty']
sales_tax = 0.0925
df['total_price']= (df['unit_price_sum']+df['shipping_price'])*(1+sales_tax)
df['total_price']= df['total_price'].round() 
df[['unit_price','unit_weight','qty','total_weight','shipping_price','total_price']].head(3)


unit_price  unit_weight  qty  total_weight  shipping_price  total_price
0     1096.80         7.50  105        787.50         5512.50     131839.0
1       24.95         1.49   21         31.29          312.90        914.0
2       13.52         1.68   39         65.52          458.64       1077.0

In [123]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['cost']=df['unit_cost']*df['qty']+df['shipping_price']
df[['cost','unit_cost','qty','shipping_price']].head(3)
#del df['subtotal']

cost  unit_cost  qty  shipping_price
0  85597.05     762.71  105         5512.50
1    629.79      15.09   21          312.90
2    765.18       7.86   39          458.64

In [126]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['cost']

df[['profit','total_price','cost']].head(3)


profit  total_price      cost
0  46241.95     131839.0  85597.05
1    284.21        914.0    629.79
2    311.82       1077.0    765.18

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [146]:
# Check your work using the totals above
order_df = [2742071,2173913,6128929]

def order_consumption_price(order_id):
    return df.loc[df['order_id'] == order_id,'total_price'].sum()

for order_id in order_df:
    total_price_found=order_consumption_price(order_id)
    print(f"Order{order_id} had a total price : $ {total_price_found:.2f}")

Order2742071 had a total price : $ 152813.00
Order2173913 had a total price : $ 162389.00
Order6128929 had a total price : $ 923441.00


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [272]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

top5_client_list

def client_consumption(top5_client_list):
    return df.loc[df['client_id'] == client,'total_price'].sum()
def client_qty_overview(top5_client_list):
    return df.loc[df['client_id'] == client,'qty'].sum()
def client_shippint_overview(top5_client_list):
    return df.loc[df['client_id'] == client,'shipping_price'].sum()
def client_cost_overview(top5_client_list):
    return df.loc[df['client_id'] == client,'cost'].sum()
def client_profit_overview(top5_client_list):
    return df.loc[df['client_id'] == client,'profit'].sum()

client_vips =[]


for client in top5_client_list:
    client_consump_overview=client_consumption(client)
    qty_vip=client_qty_overview(client)
    shipping_price_vip=client_shippint_overview(client)
    line_cost=client_cost_overview(client)
    line_profit=client_profit_overview(client)
    customized_vips = {
        'client_id':client,
        'qty':qty_vip,
        'shipping_price':shipping_price_vip,
        'line_price':client_consump_overview,
        'line_cost':line_cost,
        'line_profit':line_profit

    }
    client_vips.append(customized_vips)

    print(f"{client} : $ {client_consump_overview:.2f}")


33615 : $ 8377308.00
66037 : $ 10259518.00
46820 : $ 9743791.00
38378 : $ 12906555.00
24741 : $ 82268900.00


In [275]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
client_vips_df = pd.DataFrame(client_vips)

client_vips_df = client_vips_df.sort_values(by='line_profit')
client_vips_df
 


client_id     qty  shipping_price  line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.0   6175313.91   2201994.09
2      46820   75768      1601448.84   9743791.0   7007191.64   2736599.36
1      66037   43018      1395151.85  10259518.0   7004482.98   3255035.02
3      38378   73667      3429455.40  12906555.0   9634720.98   3271834.02
4      24741  239862      5126448.37  82268900.0  45688899.71  36580000.29

In [307]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

desc_profit_vip=client_vips_df.rename({'client_id': 'Client ID', 'qty': 'Units','shipping_price': 'Shipping', 
                 'line_price': 'Total Revenue','line_cost': 'Total Cost', 'line_profit': 'Total Profit'}, axis=1)
format_coloumns = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']
desc_profit_vip_overview = desc_profit_vip.copy()

# Apply the formatting to the specific columns
desc_profit_vip_overview[format_coloumns] = desc_profit_vip_overview[format_coloumns].apply(lambda x: x.apply(lambda val: f'${val / 1_000_000:.2f}M'))



In [306]:
# Sort the updated data by "Total Profit" form highest to lowest

desc_profit_vip_overview=desc_profit_vip_overview.sort_values(by='Total Profit', ascending=False)
desc_profit_vip_overview

Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M

In [ ]:
#Write a brief 2-3 sentence summary of your findings.
"""From this challenge to practice panda skill on client_dataset.csv, 
I learned how to sort information and find overview information of one data sheet,
As well, I practiced my analytic skill to give an overview based on specific requirment
and informations!
I also gain the skill to format the information according to requirment

"""